### ISSUE TITLE
Xamarin.Forms error binding to property where CanWrite returns false?

### ISSUE THREAD

**Issue by ajj7060** [MarimerLLC/cslaforum#708]
**Question**
I'm creating a property AuthorizationRule, where the logic looks at the state of the BO to help determine if the property may be changed.  This works fine, but I'm hitting an issue in Xamarin.Forms.  It seems like XF when doing its binding is reading all the property values, and then immediately trying to set them again (even if the value being set already matches the value in the property).  

It seems like a bug to me (in XF, not Csla), but I wanted to see if anyone here has encounter this, and if so, is there any way to get it to work correctly?

I'm trying to hide the control if the user doesn't have permission to write to it (which seems to work fine), but setting XF is triyng to set the property the app blows up because Csla throws a SecurityException (as it should).

**Version and Platform**
CSLA version: 4.7.200
OS: Windows, iOS, Android
Platform: Xamarin

**Comment by rockfordlhotka** [comment]
The standard `SetProperty` helper should not be invoking your rules if the current value and the new value are the same - it has code to short-circuit processing in that case (because Windows Forms used to (and maybe still does) send the same value to the property over and over).

However, that short-circuit _does_ happen after the authz check, so I suppose that's why you are seeing this issue.

The design philosophy behind the CSLA behavior is that if the user isn't authorized to change a property, the UI control will be set to read-only/disabled/hidden or whatever is necessary to show the user that they can't change the value.

You can use the `PropertyInfo` type (in `Csla.Xaml`) as an intermediary between the UI control and the underlying business object property. The PropertyInfo type provides a `Value` property so you can bind the value, plus other metastate properties about the property so you can do other binding.

**Comment by ajj7060** [comment]
So you're saying to bind the Entry's Text property to the Value property of the PropertyInfo, instead of to the Csla's objects property?  I will try that tomorrow.

I do have a PropertyInfo in the Xaml, which I was using to bind the Entry's IsVisible property to the PropertyInfo's CanWrite property, but XF still tries to set the BO's property (maybe before it binds IsVisible?)

**Comment by rockfordlhotka** [comment]
Yes, `PropertyInfo` is designed to be the single point of binding between a UI control and a property (and its metastate).

**Comment by ajj7060** [comment]
I tried binding the Entry's Text property to the Value on PropertyInfo, but the app still crashes with a security exception.

The page is bound to a subclass of Csla's ViewModelBase<T>, so Model, is that property from the view model.

			<Label
				IsVisible="{Binding Path=CanWrite, Source={x:Reference CustomerNamePropertyInfo}}"
				Text="Customer Name"
				FontAttributes="Bold"
				TextColor="Blue"
				HorizontalOptions="StartAndExpand" />
			<Entry
				IsVisible="{Binding Path=CanWrite, Source={x:Reference CustomerNamePropertyInfo}}"
				Text="{Binding Path=Value, Source={x:Reference CustomerNamePropertyInfo}}" />
			<csla:PropertyInfo x:Name="CustomerNamePropertyInfo" Path="Model.CustomerName"/>

**Comment by rockfordlhotka** [comment]
This may be something to be changed/fixed in `PropertyInfo` then - if we know XF has this behavior then perhaps we can short-circuit the issue at that level.

I'd really prefer to avoid changing `SetProperty` because that affects all UI platforms and the risk of sucha  change is much higher.

**Comment by Bowman74** [comment]
@ajj7060 
This is how I've bound in Xamarin.Forms with CSLA properties in the past using the propertyinfo control:

`<controls:PropertyInfo x:Name="LastNameInfo" BindingContext="{Binding User}" Path="LastName" />
 <controls:ExtendedEntry BindingContext="{x:Reference LastNameInfo}" Text="{Binding Value, Mode=TwoWay}" Placeholder="Last Name" Style="{StaticResource EntryColumnStyle}" ErrorMessage="{Binding RuleDescription}" />`

On the ViewModel it had a property called User that returned an instance of the CSLA derived user class. The User class had a property called LastName. I remember I had to mess with the binding syntax of the PropertyInfo control before I got it to work and I remember it has to be just right.

As of about a year ago, this worked for me.

**Comment by ajj7060** [comment]
@rockfordlhotka Handling it in the PropertyInfo for XF sounds like a good idea.  To be clear I wasn't asking for any changes in Csla, I was just wondering how others that may have hit this issue have handled it.  I've worked around it for now by adding a CustomerName and CanWriteCustomerName property to the VM, and binding to that.  In the VM's CustomerName setter, I just ignore the value set if CanWriteProperty on the Csla object returns false.  I'm not entirely thrilled just ignoring the databinding, but at the same time the control shouldn't be visible anyway, so I think it should work.

@Bowman74 I did try that, the issue though is that for a property where the BO denys writing, it will blow up because XF tries to set the property, even if the control is not visible (and its not actually changing the propery).

**Comment by Bowman74** [comment]
@ajj7060 Ah got it. Yes, I've noticed in the past that indeed XF does work that way. It does an unnecessary setting of the binding source with the value it just got from it. 

I agree that the only way I can think about fixing it is as you say the PropertyInfo control itself intercepted the set and didn't allow it or swallowed the exception. That would work more like value converters where I also believe swallow any exceptions that happen internally.
